transition to nucleo for fuzzy matching

pascalkuthe opened the pull request:
Closes #1707
Closes #1987
Closes #1543
Closes #5871
Closes #7645
Closes #7652

~**NOTE** This PR is currently still experimental I want to tweak the fuzzy matching scoring calculations in nucleo a bit more, add more tests, cleanup the nucleo codebase. There might also be some issues with the completers~

This PR fully replaces the skim matcher with used for all fuzzy matching in helix currently (completions, picker, ..) with `nucleo`. I have written nucleo to be as performant as possible (heavily referenced fzf). The matching algorithm itself can often be 8 times faster than skim.

https://github.com/helix-editor/nucleo

Furthermore, `nucleo` offers a high-level API that essentially replaces the entire picker logic in helix. Compared to the current helix `nucleo`s' high-level API comes with some nice features:

* matching in a background thread pool (no freezing the UI thread)
* match items can be streamed into the matches (lock-free)

I think the following demonstrates the practical use case well:

I am able to fuzzy match to the helix changelog very quickly.. From my root directory:


https://asciinema.org/a/MnRf8JGPFll1jw1GpHwNGGCXU

On the current helix version I have to wait for all (3 million) files to be crawled (with the editor frozen in the meantime) and matching is slowed to a crawl afterwards.

Just for completeness another capture of actually waiting for all files to stream in and fuzzy matching on the result (still just as fast/no noticeable lag)

https://asciinema.org/a/v1RZmq5wV80hf2BWSM1RUqaod


Nulceo is **faster than fzf** (see the comparison in the README) while yielding the same results (and even better results in some cases)

pascalkuthe commented:
Ah dann I forgot to switch from a lath to git dependency before pushing. Will fix that tomorrow

pascalkuthe commented:
I added the git dependency, I also added some further improvements upstream to nucleo (even further performance improvements and improving matching quality in a few cases). I also found/fixed the regression I had accidently introduced to the filename completer.

i want to improve some more thing upstream in nulceo in the future but those shouldn't significantly affect helix (or usage) so I think this ir ready for testing/review now. It would be nice if people could give this a spin so I can weed out any remaining upstream issues (if there are any)

gabydd commented:
getting a really weird panic when searching in my dev/home directory which has a bunch of files and then deleting the search, ascinnema attached it's not been reproducible but I wanted to report it just incase and I'll dig a bit more into it later tonight (all the panics are the same though: ``attempt to subtract with overflow``)
https://asciinema.org/a/Sc3tPpQhgpiu6gp4J2ujpe1Jc
<details>
<summary>Details</summary>

```
thread 'nucleo worker 0' panicked at 'attempt to subtract with overflow', /home/gaby/.cargo/git/checkouts/nucleo-fe29e1ee969779b0/f4e19b4/src/worker.rs:244:27
stack backtrace:
   0:     0x55eb0ee94e20 - std::backtrace_rs::backtrace::libunwind::trace::h32eb3e08e874dd27
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55eb0ee94e20 - std::backtrace_rs::backtrace::trace_unsynchronized::haa3f451d27bc11a5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55eb0ee94e20 - std::sys_common::backtrace::_print_fmt::h5b94a01bb4289bb5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x55eb0ee94e20 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hb070b7fa7e3175df
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x55eb0eebf51e - core::fmt::write::hd5207aebbb9a86e9
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/fmt/mod.rs:1202:17
   5:     0x55eb0ee8d665 - std::io::Write::write_fmt::h3bd699bbd129ab8a
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/io/mod.rs:1679:15
   6:     0x55eb0ee96b83 - std::sys_common::backtrace::_print::h7a21be552fdf58da
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x55eb0ee96b83 - std::sys_common::backtrace::print::ha85c41fe4dd80b13
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x55eb0ee96b83 - std::panicking::default_hook::{{closure}}::h04cca40023d0eeca
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:295:22
   9:     0x55eb0ee9686f - std::panicking::default_hook::haa3ca8c310ed5402
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:314:9
  10:     0x55eb0cd7fbd0 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h4d50a41a4cddb9fb
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1954:9
  11:     0x55eb0cdc7cc6 - helix_term::application::Application::run::{{closure}}::{{closure}}::h423bcb8900f006d5
                               at /home/gaby/dev/helix/helix-term/src/application.rs:1183:13
  12:     0x55eb0ee9732d - std::panicking::rust_panic_with_hook::h7b190ce1a948faac
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:702:17
  13:     0x55eb0ee97141 - std::panicking::begin_panic_handler::{{closure}}::hbafbfdc3e1b97f68
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:586:13
  14:     0x55eb0ee952cc - std::sys_common::backtrace::__rust_end_short_backtrace::hda93e5fef243b4c0
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:138:18
  15:     0x55eb0ee96ea2 - rust_begin_unwind
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
  16:     0x55eb0eebd253 - core::panicking::panic_fmt::h8d17ca1073d9a733
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
  17:     0x55eb0eebd09d - core::panicking::panic::hf0565452d0d0936c
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:48:5
  18:     0x55eb0d159635 - nucleo::worker::Woker<T>::run::hc3ed853105efd098
                               at /home/gaby/.cargo/git/checkouts/nucleo-fe29e1ee969779b0/f4e19b4/src/worker.rs:244:27
  19:     0x55eb0d223563 - nucleo::Nucleo<T>::tick_inner::{{closure}}::h1a38f7f6b214b6e2
                               at /home/gaby/.cargo/git/checkouts/nucleo-fe29e1ee969779b0/f4e19b4/src/lib.rs:213:41
  20:     0x55eb0d30a600 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hed3b42b07acd491e
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panic/unwind_safe.rs:271:9
  21:     0x55eb0d4405df - std::panicking::try::do_call::hae5637ca2cbf4ebf
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:492:40
  22:     0x55eb0d455bcb - __rust_try
  23:     0x55eb0d42dc96 - std::panicking::try::hd3a28a4bf1f782bd
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:456:19
  24:     0x55eb0d115e10 - std::panic::catch_unwind::h69199f86625d822d
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panic.rs:137:14
  25:     0x55eb0d3659c4 - rayon_core::unwind::halt_unwinding::h1062f2601e7722ab
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/unwind.rs:17:5
  26:     0x55eb0d582ea6 - rayon_core::registry::Registry::catch_unwind::h1f744e432843e34a
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/registry.rs:376:27
  27:     0x55eb0d4641f9 - rayon_core::spawn::spawn_job::{{closure}}::h9fac78758bf53797
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/spawn/mod.rs:97:13
  28:     0x55eb0d06d84d - <rayon_core::job::HeapJob<BODY> as rayon_core::job::Job>::execute::h6370fc3fd8760662
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/job.rs:169:9
  29:     0x55eb0e709200 - rayon_core::job::JobRef::execute::hf2f968d4685f9327
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/job.rs:64:9
  30:     0x55eb0e713350 - rayon_core::registry::WorkerThread::execute::h72aa01593b2acdf7
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/registry.rs:874:9
  31:     0x55eb0e713061 - rayon_core::registry::WorkerThread::wait_until_cold::h6d93b93b6c9047e0
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/registry.rs:820:17
  32:     0x55eb0e712e5c - rayon_core::registry::WorkerThread::wait_until::ha9245a91ace4c8a4
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/registry.rs:803:13
  33:     0x55eb0e713943 - rayon_core::registry::main_loop::hd7edcc9863c2d994
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/registry.rs:948:5
  34:     0x55eb0e70e9bc - rayon_core::registry::ThreadBuilder::run::h9f5ff30c97a82c04
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/registry.rs:54:18
  35:     0x55eb0e70ee6d - <rayon_core::registry::DefaultSpawn as rayon_core::registry::ThreadSpawn>::spawn::{{closure}}::h317247fb0edde1a1
                               at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.11.0/src/registry.rs:99:20
  36:     0x55eb0e717c3f - std::sys_common::backtrace::__rust_begin_short_backtrace::h511fa934ef884872
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:122:18
  37:     0x55eb0e6f4f5d - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::h4e17354ecce056a1
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/mod.rs:514:17
  38:     0x55eb0e6e0b91 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h62e25f069ce3ee35
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panic/unwind_safe.rs:271:9
  39:     0x55eb0e700a3a - std::panicking::try::do_call::h84c99c20194314e5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:492:40
  40:     0x55eb0e70318b - __rust_try
  41:     0x55eb0e70053c - std::panicking::try::h3cf32ffee9d08df9
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:456:19
  42:     0x55eb0e706601 - std::panic::catch_unwind::h7bac15597f9176b8
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panic.rs:137:14
  43:     0x55eb0e6f496b - std::thread::Builder::spawn_unchecked_::{{closure}}::h967d70cb241066ec
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/thread/mod.rs:513:30
  44:     0x55eb0e71e8af - core::ops::function::FnOnce::call_once{{vtable.shim}}::h3e97e6b290b31b45
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:248:5
  45:     0x55eb0ee9d1f3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h49f797984e2121bf
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1940:9
  46:     0x55eb0ee9d1f3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hfa4f3d0ee6440e0b
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/alloc/src/boxed.rs:1940:9
  47:     0x55eb0ee9d1f3 - std::sys::unix::thread::Thread::new::thread_start::h62ca48b42d48a8fc
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys/unix/thread.rs:108:17
  48:     0x7f07a30889eb - <unknown>
  49:     0x7f07a310d23c - <unknown>
  50:                0x0 - <unknown>
Rayon: detected unexpected panic; aborting
fish: Job 1, 'RUST_BACKTRACE=full ./helix/tar…' terminated by signal SIGABRT (Abort)
```

</details>

gabydd commented:
Also does this fix https://github.com/helix-editor/helix/issues/7645 as well?

pascalkuthe commented:
Thanks! I probably missed this earlier because I was running release builds (where this silently wraps which shoulst really cause too many problems at that particular subtraction. I will try to reproduce this. Might need to aetyo a fuzzer for the high level API tough since that's probably hard to reproduce 😅


I will see if I can find a quick fix this tomorrow morining, I will be travelling the rest of the weekend so it make take a while if I don't find it quickly

pascalkuthe commented:
> Also does this fix #7645 as well?

Yeah I forgot about that 👍

pascalkuthe commented:
> getting a really weird panic when searching in my dev/home directory which has a bunch of files and then deleting the search, ascinnema attached it's not been reproducible but I wanted to report it just incase and I'll dig a bit more into it later tonight (all the panics are the same though: `attempt to subtract with overflow`) [asciinema.org/a/Sc3tPpQhgpiu6gp4J2ujpe1Jc](https://asciinema.org/a/Sc3tPpQhgpiu6gp4J2ujpe1Jc)
> full backtrace

I found and fixed the bug, it was a pretty easy to fix oversight

archseer commented:
Let's get a couple days of testing in before this gets merged, I'm going to go ahead and test run this branch

archseer commented:
The file list has a fair bit of shuffling around as more files are scanned, it seems like the new files are added at the top of the list then sorted. So the files with the top score move slightly downwards, then are moved back up the list.

pascalkuthe commented:
> Is there also some kind of total limit? I can't see a file tree if I open it in `/nix/store`, it just shows 0 files. Maybe this already happens on master though

Hmm in there is no hard limit but the nix store may be hidden by some filter?

pascalkuthe commented:
> The file list has a fair bit of shuffling around as more files are scanned, it seems like the new files are added at the top of the list then sorted. So the files with the top score move slightly downwards, then are moved back up the list.

Hmm I noticed that too tgis moringing. But only in debug mode weirdly enough not in release mode but I did want to investigate that futrher. I do actually was pretty careful ti ensure this doesn'happen so en I am bit surprised where this is coming from. I am travelling this weekend so I didn't have time ti investigate further, will take a closer look next week.

archseer commented:
I'm always building in release mode but I needed to execute on a very large tree so that the file list would take some time to load. I searched for `test` under `~` and I have about 500k files

archseer commented:
> We do some filters around symlinks so it may be related to that I guess or you may have an ignore file somewhere?

Ah right, I think `/nix/store` is exclusively symlinks

pascalkuthe commented:
> 

Hmm I didn't really have the issue when I tried it (streaming about 3 million files from my root directory, I searched fro changelog so there are quite a few matches streaming in). 

I will try to test on my laptop later. Maybe I can find a quick fix. Nucleo should always resort if there are new matches (in fact I have a dedicated flag in the matcher that controls where the match snapshot will be updated and that is only set to true if sorting happened) so it's hopefully just a small oversight

gabydd commented:
i was getting this as well,  it was easier to see when I was in my home or root directory and I turned off not showing hidden files (``toggle file-picker.hidden``)  searching for something where a full match would be a couple folders deep so there are some matches for directories or files named helix but also for places where the helix letters are spread out, it will start cycling between the same couple files. You can see it here a couple seconds into to matching it starts shuffling a lot: https://asciinema.org/a/Z0OyFjCcrHC6hkjFzE2D2osJd

pascalkuthe commented:
While I am using an unstable (parallel) sort that shouldn't theoretically cause these problems because I am sorting by `(score, Len, idx)` and atleast `idx` should always be different. 

I suspect that the same items endsup in the list of matches (which is then sorted) twice somehow during streaming.

gabydd commented:
ah seems like I found a bit of a memory leak after leaving open the file picker in the root directory. Steps look something like this 
1. open helix in the root directory (helix finds about 400,000 files but it changes every time which probably adds to your above idea)
2. open the file picker
3. open up something like ``htop``
4. watch as helix memory goes up even once it's done running, also still using 100% cpu on one core

pascalkuthe commented:
> 

I think that's mostly the file crawler itself and not nucelo itself. If the file crawler freezes then the items won't be freed. It's likely that your file crawler freezes because it fails to open some directory (like a block device or similar). I could force the deallocation by using a weak reference but that would add quite a bit of overhead to the file walker. Ideally the file walker wouldn't freeze but I amnnot sure if there is a way to prevent that.

.this us also likely the cause of finding a different number of items each time. The file crawler freezes at some specific file and has crawled a different number of files before that

pascalkuthe commented:
I think using the weak reference might actually be ok but I am not sure if I can fix the single cores sitting at 100% memory consumption for edge cases like this. Ideally we would filter these files somehow (likely block devices or similar) but that is or scope for this PR IMO

gabydd commented:
ah that makes a lot of sense, yeah I think a core sitting at 100% is probably fine, thanks for looking into it

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -2566,7 +2568,7 @@ fn file_picker(cx: &mut Context) {
         return;
     }
-    let picker = ui::file_picker(root, &cx.editor.config());
+    let picker = ui::file_picker(root, &cx.editor.config(), cx.editor.redraw_handle.0.clone());
```
What do you think about adding a small `#[inline]` function to Editor to do the same as this? I'm thinking that if we refactor RedrawHandle in the future (for example from a tuple to a struct) we might end up having to change a lot of callsites. I'm not sure what a good name for this function would be.

pascalkuthe replied:
Yeah that's probably a good idea. An alternative I have been thinking about is to make the redraw handle a static. A `Notify` can be const initialized (and never needs mutable access) so this would avoid having to add the extra argument everywhere.

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -190,55 +197,51 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> Picker
+    let injector = picker.injector();
+    std::thread::spawn(move || {
+        for file in files {
+            if injector.push(file) {
```
The `true` return meaning shutdown seems unintuitive to me. Could this be refactored as an enum or a Result?

pascalkuthe replied:
Hnn yeah result is probably cleaner and close to his channels do it.

the-mikedavis commented on helix-term/src/ui/picker.rs:
```diff
@@ -202,122 +299,25 @@ impl<T: Item + 'static> Picker<T> {
     ) -> Self {
         self.file_fn = Some(Box::new(preview_fn));
+        // assumption: if we have a preview we are matching paths... If this is ever
+        // not true this could be a seperate builder function
```
```suggestion
        // not true this could be a separate builder function
```

the-mikedavis commented on helix-term/src/ui/picker.rs:
```diff
@@ -539,106 +557,90 @@ impl<T: Item + 'static> Picker<T> {
+        let mut indices = Vec::new();
+        let mut matcher_lock = MATCHER.lock();
+        let matcher = matcher_lock.get_or_insert_with(nucleo::Matcher::default);
+        // TODO: paths
+        matcher.config = MatcherConfig::DEFAULT;
+        if self.file_fn.is_some() {
+            matcher.config.set_match_paths()
+        }
```
Is this TODO resolved (at least temporarily) by L569-571?

pascalkuthe replied:
Yeah forgot to remove that 😅

the-mikedavis reviewed (commented):
I have a few minor typo fixes and some comments/questions but overall the changes look really good, excellent work! I'm surprised the diff isn't larger given the big change in behavior.

pascalkuthe commented on helix-core/Cargo.toml:
```diff
@@ -48,6 +48,9 @@ chrono = { version = "0.4", default-features = false, features = ["alloc", "std"
 etcetera = "0.8"
 textwrap = "0.16.0"
 
+nucleo = { version = "0.1", git = "https://github.com/helix-editor/nucleo.git" }
```
Yeah that was my plan 👍 I wanted to wait till we are essentialky ready to emrge sk K dont havr tk create a bunch if bugfix releases. I want to look into a fix for the jittering behavior described above

pascalkuthe commented on helix-core/src/fuzzy.rs:
```diff
@@ -0,0 +1,22 @@
+use nucleo::{CaseMatching, MatcherConfig};
+use parking_lot::Mutex;
+
+pub static MATCHER: Mutex<Option<nucleo::Matcher>> = Mutex::new(None);
```
Yeah I like that idea 👍 It felt a little ugly having to reinit the marcher in `helix-temr`

pascalkuthe replied:
I wound up doing something similar but slightly different. I kept the API closer to a mutex (and called the type `LayzMutex`) and used a mapped mutex lock guard instead of a `with` method. That allowed me to avoid nesting a bunch of code into closures.

gabydd commented:
investigated the memory leak it was proc causing the problems, but it also turns out this is somewhat a regression in rust that was fixed for 1.68 (tested changing the rust-toolchain.toml to 1.68, rebuilding and searching in root and there is no memory leak or 100% cpu core) see the relevant discussion in https://github.com/sharkdp/fd/issues/288 and https://github.com/rust-lang/rust/pull/105638 I guess firefox msrv isn't 1.68 yet but do we still want to upgrade to use this fix?

gabydd commented:
ah two other things, the file previews in global search don't get highlighted anymore and this might be a good change but now when you start filtering it keeps your position in the picker whereas before it moved you back to the top

gabydd commented:
trying to go through all the changed pickers found that the workspace symbol search reliably panics once you type two characters into it, opening it from any file in helix. I used ``helix-core/src/comment.rs`` and searched for ``co`` then you get this backtrace:

<details>
<summary>Full Backtrace</summary>

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', helix-term/src/ui/picker.rs:574:53
stack backtrace:
   0:     0x557573287afe - std::backtrace_rs::backtrace::libunwind::trace::h595f06c70adcc478
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x557573287afe - std::backtrace_rs::backtrace::trace_unsynchronized::h177a0149c76cdde9
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x557573287afe - std::sys_common::backtrace::_print_fmt::hc0701fd2c3530c58
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x557573287afe - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd4cd115d8750fd6c
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x5575726f094e - core::fmt::write::h93e2f5923c7eca08
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/core/src/fmt/mod.rs:1213:17
   5:     0x557573280e95 - std::io::Write::write_fmt::h8162dbb45f0b9e62
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/io/mod.rs:1682:15
   6:     0x5575732878d5 - std::sys_common::backtrace::_print::h1835ef8a8f9066da
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x5575732878d5 - std::sys_common::backtrace::print::hcb5e6388b9235f41
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x55757328979f - std::panicking::default_hook::{{closure}}::h9c084969ccf9a722
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:267:22
   9:     0x557573289514 - std::panicking::default_hook::h68fa2ba3c3c6c12f
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:286:9
  10:     0x557573289eb1 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h4e6ced11e07d8b24
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/alloc/src/boxed.rs:2002:9
  11:     0x557573289eb1 - std::panicking::rust_panic_with_hook::h8d5c434518ef298c
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:692:13
  12:     0x557573289c15 - std::panicking::begin_panic_handler::{{closure}}::hf33414f5dabf6faf
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:577:13
  13:     0x55757328802c - std::sys_common::backtrace::__rust_end_short_backtrace::hc50389427413bb75
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/sys_common/backtrace.rs:137:18
  14:     0x557573289982 - rust_begin_unwind
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:575:5
  15:     0x5575726ee6f3 - core::panicking::panic_fmt::h2de7a7938f816de8
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/core/src/panicking.rs:64:14
  16:     0x5575726ee78d - core::panicking::panic::h90931f06a97cc5e0
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/core/src/panicking.rs:114:5
  17:     0x557572e76597 - <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold::h18bd8e60e73ed595
  18:     0x557572e228da - <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter::h863989836c6a0e89
  19:     0x557572dc1972 - <helix_term::ui::picker::Picker<T> as helix_term::compositor::Component>::render::h785544d66f79f5d1
  20:     0x55757305dd43 - helix_term::application::Application::render::{{closure}}::h318da8f3b1e8e937
  21:     0x55757305b6fd - helix_term::application::Application::run::{{closure}}::h67f9e18848b684b0
  22:     0x55757307154d - tokio::runtime::park::CachedParkThread::block_on::h268c55b4e22d9248
  23:     0x557573051ab7 - tokio::runtime::context::runtime::enter_runtime::h6f1e9f61695462dd
  24:     0x5575730ab919 - tokio::runtime::runtime::Runtime::block_on::h8187e65563576e99
  25:     0x5575730ba08a - hx::main::h0b945d8dbf578c46
  26:     0x557573079153 - std::sys_common::backtrace::__rust_begin_short_backtrace::hdaac3b2e7928deed
  27:     0x5575730a247d - std::rt::lang_start::{{closure}}::ha3839f340e000a7d
  28:     0x557573279b38 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h699977d052768608
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/core/src/ops/function.rs:287:13
  29:     0x557573279b38 - std::panicking::try::do_call::h4e121e623c70f903
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:483:40
  30:     0x557573279b38 - std::panicking::try::hf9d919e062bc178a
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:447:19
  31:     0x557573279b38 - std::panic::catch_unwind::h7a7b12272684cb97
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panic.rs:140:14
  32:     0x557573279b38 - std::rt::lang_start_internal::{{closure}}::hd96b0eb4844b8762
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/rt.rs:148:48
  33:     0x557573279b38 - std::panicking::try::do_call::h1af1f88f4f92a22c
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:483:40
  34:     0x557573279b38 - std::panicking::try::hf20d7abea7f0f097
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panicking.rs:447:19
  35:     0x557573279b38 - std::panic::catch_unwind::hb0e084c3a9c042e4
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/panic.rs:140:14
  36:     0x557573279b38 - std::rt::lang_start_internal::hca9d5c7277f5b67c
                               at /rustc/2c8cc343237b8f7d5a3c3703e3a87f2eb2c54a74/library/std/src/rt.rs:148:20
  37:     0x5575730ba185 - main
  38:     0x7f7c67223cd0 - <unknown>
  39:     0x7f7c67223d8a - __libc_start_main
  40:     0x55757266cb35 - _start
  41:                0x0 - <unknown>
```

</details>

I think I have gone through most of the pickers now but it's still good to have multiple eyes on this to make sure everything is covered

pascalkuthe commented:
> investigated the memory leak it was proc causing the problems, but it also turns out this is somewhat a regression in rust that was fixed for 1.68 (tested changing the rust-toolchain.toml to 1.68, rebuilding and searching in root and there is no memory leak or 100% cpu core) see the relevant discussion in [sharkdp/fd#288](https://github.com/sharkdp/fd/issues/288) and [rust-lang/rust#105638](https://github.com/rust-lang/rust/pull/105638) I guess firefox msrv isn't 1.68 yet but do we still want to upgrade to use this fix?

Thanks for looking into this! I suspected something like this but was not able to find anything before. 

I would be in favor  of changing the rust toolchain defined in the repo to 1.68 without bumping the MSRV. I think this bug is serious enough to warrant that. We can safely bump our MSRV to 1.66 (just forgot about that so far). There was very little stabilized in 1.67/1.68 so its not too inconvenient, we can add an extra CI check (like we had in the past)  until we bump our MSRV to 1.68. This should be a seperate PR IMO since its orthogonal/an existing issue. Thaughts @the-mikedavis @archseer?

pascalkuthe commented:
> ah two other things, the file previews in global search don't get highlighted anymore and this might be a good change but now when you start filtering it keeps your position in the picker whereas before it moved you back to the top

I will look into the highlight. That one is odd to me since nothing should have changed there.

The second point is intentional. This is how fyf/skim behave too. The reason for the change is the streaming behavior. If we reset the cursor to the top each time the matches change that would mean the cursor would be stuck at the top all the time while streaming. So I opted to consistently change the cursor to avoid jumping

zetashift commented:
Curiosity question: does this have any impact on https://github.com/helix-editor/helix/pull/7265 ?

pascalkuthe commented:
I have fkubd fixes for the issues in nucleo that @archseer and @gabydd found (see helix-editor/nucleo#5).

For the sorting: I was just [being stupid](https://github.com/helix-editor/nucleo/pull/5/commits/d500fb90dddb28345c71edd0721ef35a4f10f2ff). This was a new regression si that's why I rember it working on the past.

For the crash@gabydd found it was si ply a matter of consistently treating the Matcher results as snapshots (which may refer to different item arrays). I tweaked the API if nucleo to reflect that in the type system (also I forgot a pretty trivial update of a variable which made the crash way more common but the second change was still needed to fix all cases).

I will update this branch tomorrow once I merge that PR into nucleo.

doy commented:
hmmm, the multi-word behavior is still not great:
![nucleo](https://github.com/helix-editor/helix/assets/78547/93ed628c-9318-41f2-a501-79cf42ca3d90)
![nucleo2](https://github.com/helix-editor/helix/assets/78547/c6e248a3-c3d0-45de-bb89-415548a3fc6a)
it seems slightly better than the skim results, but still not as good as fzf - i would expect more weight to be placed on consecutive characters

pascalkuthe commented:
The sorting is currently broken as I mentioned above ai anything NJT the first match is basically random. Probably fixed by the patch I linked

pascalkuthe commented:
![image](https://github.com/helix-editor/helix/assets/61850714/74994ecb-f691-4a1c-bffc-e617722bd1f0)

yeah this is fixed with the latest changes I think (see screenshot, from the filenames I assumed you were matching in rust-lang/rust)

pascalkuthe commented:
> Curiosity question: does this have any impact on #7265 ?

nucleo actually has columnar matching builtin first class in the high level API, that plus the streaming should actually help a bit for that PR

pascalkuthe commented:
I think I got most things. I am leaning towards introducing static (https://github.com/helix-editor/helix/pull/7814#discussion_r1285022995) but waiting for feedback on my idea, of-course the crates.io publish also needs to happen but I want to do that last to avoid publishing multiple versions.

The only thing I wasn't able to address at all was the missing highlight in the global search picker reported by@gabydd. I wasn't able to reproduce that. Could you provide a reproduction case (or a screenshot comparing master/this branch to show the expected/actual behavior)?

gabydd commented:
while I was testing this made me notice something else so the recordings who two things the first thing is the theme menu doesn't popup until you start typing a theme where on master it just instantly pops up, and the second part is global searching in the helix repository for ``ignore`` and you can see how it only highlights files that are already open where it should highlight files on idle time-out oh but I just realized this might be similar to https://github.com/the-mikedavis/helix/commit/24916dc0dc7afc68a9d286f9ce2546a3b5bce534 but instead of it being a dynamic picker you are doing:
```rs
compositor.push(Box::new(picker))
```
instead of 
```
https://asciinema.org/a/N3o9tgeVL2RfNEi5Rsvr0Qskm
https://asciinema.org/a/5BOvBzOicGRod891jGeeNcan6

pascalkuthe commented:
Ah I see yeah that's pretty subtle thanks for digging into that

gabydd commented:
the issue with the theme command is the same for all the other completers that use  ``fuzzy_match`` when there is an empty input, except for ``filename_impl`` because it doesn't fuzzy match when there is empty input, I am guessing fuzzy_match isn't returning any results when it get's an empty string

pascalkuthe commented:
I fixed the issues @gabydd found

pascalkuthe commented:
I addressed the last comment by @the-mikedavis by moving the redraw handle to its own create `helix-event`. The crate is currently pretty empty  but will hopefully soon be filled by a generic hook/event system that I have in the works anyway (and is mostly finished). Moving the redraw handle there removes the need to modify the call-site of every picker.

It also hides the implementation details of the redraw handle from the rest of the editor and reduced complexity in a bunch of cases. This would have reduced the diff a bunch but I added quite a bit of docs (even tough this is technically existing code) and adding a new crate also and the diff look a bit larger. But overall I am very pleased with this change and think it made the redraw handle a lot cleaner/less hacky.

But if you disagree @archseer I can always rollback that commit.

gabydd commented on helix-event/src/redraw.rs:
```diff
@@ -0,0 +1,51 @@
+use parking_lot::{RwLock, RwLockReadGuard};
+use tokio::sync::Notify;
+
+/// A `Notify` instance that can be used to (asyronously) request
```
```suggestion
/// A `Notify` instance that can be used to (asynchronously) request
```

pascalkuthe replied:
I forced pushed this since I was fixing it anyway

gabydd commented on helix-event/src/redraw.rs:
```diff
@@ -0,0 +1,51 @@
+
+/// Wait until all locks acquired with [`lock_frame`] have
+/// have been released. This function is called before rendering
+/// and is intended to allow the frame to wait for async
+/// the next frame from being drawn until the returned guard
+/// is dropped.
```
I think this is missing a word after async, not sure what though cause I am bit confused by the wording

pascalkuthe replied:
Yeah I had a brain fart there 😅

gabydd reviewed (commented):
a couple grammar and spelling nits but I really like the changes, makes the ``Notify`` use much more readable also excited for what a proper event system can unlock

archseer commented on Cargo.lock:
```diff
@@ -1842,6 +1924,28 @@ dependencies = [
  "getrandom",
 ]
 
+[[package]]
+name = "rayon"
+version = "1.7.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "1d2df5196e37bcc87abebc0053e20787d73847bb33134a69841207dd0a47f03b"
+dependencies = [
+ "either",
+ "rayon-core",
+]
+
+[[package]]
+name = "rayon-core"
+version = "1.11.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "4b8f95bd6966f5c87776639160a66bd8ab9895d9d4ab01ddba9fc60661aebe8d"
+dependencies = [
+ "crossbeam-channel",
+ "crossbeam-deque",
+ "crossbeam-utils",
+ "num_cpus",
+]
```
Hmm do we even need rayon? I imagine performance would be good enough even without parallel sort

pascalkuthe replied:
I use rayon in lot int the parallel worker. It's not just the parallel sort but more importantly `par_extend' and a parallel `for_each` and in general parallel iterators (basically running the matcher in parallel). I went down the rabbit hole of replacing rayon a while ago. 

For high-performance parallel code with small units of work like this, I would say rayon is unbeatable. The work-stealing thread pool from rayon core is amazingly performant and well thought out. The parallel iterators are built around dynamically taking advantage of work splitting and are really good at assigning optimal units of work to each thread. Nucleo benefits from that a lot because the matcher has a lot of fast paths built in (like fast rejection) so it's very hard to predict how long each match will take (so if you split the list of matches into N chunks and give N thread one chunk its quite likely that one thread wil&take significantly longer). A good adaptive work-splitting approach like in rayon is just the right thing and benefits nulco a lot.

I looked into just using `rayon-core` and reimplementing just the parts of rayon that I would need but that involves copying a *lot* of (unsafe) code and doesn't feel like a good tradeoff to me.  If you feel uncomfortable pulling in all of rayon I could do that if you see it as a requirement. I would still need 'rayon-core` tough (the work-stealing worker is really good at what it does).

archseer commented on helix-term/src/application.rs:
```diff
@@ -257,16 +257,8 @@ impl Application {
-        // Acquire mutable access to the redraw_handle lock
-        // to ensure that there are no tasks running that want to block rendering
-        drop(cx.editor.redraw_handle.1.write().await);
+        helix_event::start_frame();
```
Much clearer 👍🏻

gabydd commented on helix-event/src/redraw.rs:
```diff
@@ -0,0 +1,49 @@
+    REDRAW_NOTIFY.notified()
+}
+
+/// Wait until all locks acquired with [`lock_frame`] have have been released.
```
```suggestion
/// Wait until all locks acquired with [`lock_frame`] have been released.
```

shahamran commented:
Hi @pascalkuthe, thanks a lot for this PR!
I compiled and ran this branch on my company's huge monorepo, and it panicked when searching for a specific file I knew existed.
Any idea why?

Backtrace:
```console
$ RUST_BACKTRACE=full hx
thread 'main' panicked at 'exceeded maximum length', /homes/ransh/.cargo/git/checkouts/nucleo-fe29e1ee969779b0/1756e47/src/boxcar.rs:544:23
stack backtrace:
   0:     0x555556218aa5 - std::backtrace_rs::backtrace::libunwind::trace::h32eb3e08e874dd27
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x555556218aa5 - std::backtrace_rs::backtrace::trace_unsynchronized::haa3f451d27bc11a5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x555556218aa5 - std::sys_common::backtrace::_print_fmt::h5b94a01bb4289bb5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x555556218aa5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hb070b7fa7e3175df
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x55555572d9be - core::fmt::write::hd5207aebbb9a86e9
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/fmt/mod.rs:1202:17
   5:     0x555556212425 - std::io::Write::write_fmt::h3bd699bbd129ab8a
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/io/mod.rs:1679:15
   6:     0x55555621a533 - std::sys_common::backtrace::_print::h7a21be552fdf58da
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x55555621a533 - std::sys_common::backtrace::print::ha85c41fe4dd80b13
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x55555621a533 - std::panicking::default_hook::{{closure}}::h04cca40023d0eeca
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:295:22
   9:     0x55555621a23f - std::panicking::default_hook::haa3ca8c310ed5402
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:314:9
  10:     0x55555621abff - std::panicking::rust_panic_with_hook::h7b190ce1a948faac
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:702:17
  11:     0x55555621aa64 - std::panicking::begin_panic_handler::{{closure}}::hbafbfdc3e1b97f68
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:588:13
  12:     0x555556218fac - std::sys_common::backtrace::__rust_end_short_backtrace::hda93e5fef243b4c0
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x55555621a7b2 - rust_begin_unwind
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
  14:     0x55555565ddb3 - core::panicking::panic_fmt::h8d17ca1073d9a733
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
  15:     0x55555572bb01 - core::panicking::panic_display::he956d1fbfe1d4c76
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:72:5
  16:     0x55555572baab - core::panicking::panic_str::h7d5968eea59d06d2
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:56:5
  17:     0x55555565dc26 - core::option::expect_failed::h03d502d0855ade10
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/option.rs:1880:5
  18:     0x555555d452a4 - <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold::h6f7bcfa759cb7b05
  19:     0x555555e048e6 - <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter::h8b688a69a13e747d
  20:     0x555555e37ebe - <helix_term::ui::picker::Picker<T> as helix_term::compositor::Component>::render::h2f34ae9e571db62b
  21:     0x555556046e75 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h143f9cec77e61700
  22:     0x55555604a36d - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h49f435924259001a
  23:     0x555555ff1798 - tokio::runtime::park::CachedParkThread::block_on::hc8fe2014a43c0fea
  24:     0x555555ff2497 - tokio::runtime::context::runtime::enter_runtime::hf2cb9f69f190368b
  25:     0x55555602c56b - tokio::runtime::runtime::Runtime::block_on::h4595be7470769a43
  26:     0x5555560500aa - hx::main::h5fb71060aaa35614
  27:     0x555555ff6b13 - std::sys_common::backtrace::__rust_begin_short_backtrace::h301280aa177d85f9
  28:     0x555556023639 - std::rt::lang_start::{{closure}}::h39a9e1b81c7c5419
  29:     0x55555620be3a - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hb69be6e0857c6cfb
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:283:13
  30:     0x55555620be3a - std::panicking::try::do_call::h396dfc441ee9c786
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:492:40
  31:     0x55555620be3a - std::panicking::try::h6cdda972d28b3a4f
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:456:19
  32:     0x55555620be3a - std::panic::catch_unwind::h376039ec264e8ef9
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panic.rs:137:14
  33:     0x55555620be3a - std::rt::lang_start_internal::{{closure}}::hc94720ca3d4cb727
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/rt.rs:148:48
  34:     0x55555620be3a - std::panicking::try::do_call::h2422fb95933fa2d5
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:492:40
  35:     0x55555620be3a - std::panicking::try::h488286b5ec8333ff
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:456:19
  36:     0x55555620be3a - std::panic::catch_unwind::h81636549836d2a25
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panic.rs:137:14
  37:     0x55555620be3a - std::rt::lang_start_internal::h6ba1bb743c1e9df9
                               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/rt.rs:148:20
  38:     0x5555560501a8 - main
  39:     0x7ffff6e22c87 - __libc_start_main
                               at /build/glibc-CVJwZb/glibc-2.27/csu/../csu/libc-start.c:310
  40:     0x5555556aee9a - _start
  41:                0x0 - <unknown>
```

pascalkuthe commented:
Ah i pushed an optimization to nucleo yesterday and missed an edgecase in the process which caused that crash. Should be fixed with the latest push (also rebase on master to resolve conflicts)

poliorcetics commented:
tried this branch out today, no problem and greatly improved perfs 🎉

zetashift commented:
Tested this out today, I had no issues with it and fuzzy finding something in my projects went really well...

I was searching from my `$HOME` dir! I have no idea what to do with such power, but this is very nicely done!

poliorcetics commented:
> tried this branch out today, no problem and greatly improved perfs 🎉

Ran this for the whole week, still only positive results 😄

pascalkuthe commented:
rebased to resolve conflicts with #5244

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -174,6 +179,8 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> Picker
         .git_ignore(config.file_picker.git_ignore)
         .git_global(config.file_picker.git_global)
         .git_exclude(config.file_picker.git_exclude)
+        // TODO: better sorting (anpahnumeric)
```
Looks like this colud still be resolved

the-mikedavis reviewed (commented):
This is a nice cleanup with the new changes to the redraw handle! Just a minor typo and it looks like this needs a rebase on top of https://github.com/helix-editor/helix/pull/7998 but otherwise I think this is good to go 🚀

pascalkuthe commented:
I resolved the conflicts, for the typo I simply opted to remove that TODO since I recently noticed that alphanumeric sorting isn't actually useful for a fuzzy matcher that prefers shorter matches (like nulceo). We already have an issue for this so the TODO isn't really necessary here

gabydd commented:
On the topic of alphanumeric sorting, there is a bit of a change from master in the initial sorting of the command and theme completion but it's not very significant because it will be similar or maybe better once you start filtering.

pascalkuthe commented:
hmm it's not intentional I think, could you describe more closely what changed?

gabydd commented:
Before this pr the theme completions where initially sorted fully alphanumericly so all the themes with multiple versions would be right next to each other, now they are sorted by length initially, it's just something I noticed but I don't think it's a significant change in my opinion

pascalkuthe commented:
Hmm you are right that is pretty odd since we do still sort there and while the picker does sort by length the completion matcher doesn't actually do that (and in particular it should actually do nothing if there is no filter at all)

pascalkuthe commented:
I found and fixed the bug in nucleo that's causing this, it was a small oversight/not really intended to work this way. I will publish a patch release

pascalkuthe commented:
(bumped to new version, should be fixed now)

kirawi commented:
I'm trying this out today after pulling from master, and it looks like the autocomplete sorting is not fixed.

pascalkuthe commented:
> I'm trying this out today after pulling from master, and it looks like the autocomplete sorting is not fixed.
> 
> Edit: Strangely it pinned to 0.1.0 instead of 0.1.1. I deleted the entries and tried again; works as expected now.

yeah, I updated the Cargo.lock file and thought I had committed/force-pushed the changes but it seems I forgot to `git add` before `gif commit --amend` and force pushed nothing (facepalm) I pushed the updated lockfile

archseer reviewed (commented):
Let's bump this to 0.2 then LGTM!

shahamran commented:
Hi, been testing this on a pretty big monorepo I work on. Overall the experience is tremendously better than before, but I still get relatively bad performance (probably due to being in an NFS mount).
I also have a small improvement suggestion: stream matches from directories of currently open buffers before ones from the project root, as these are more likely to be "requested".
I don't know if this was considered, and I apologize if this is not the right place to suggest it.

pascalkuthe commented:
> Let's bump this to 0.2 then LGTM!

I have released 0.2.0 and bumped this PR, this should be ready now :)



> Hi, been testing this on a pretty big monorepo I work on. Overall the experience is tremendously better than before, but I still get relatively bad performance (probably due to being in an NFS mount). I also have a small improvement suggestion: stream matches from directories of currently open buffers before ones from the project root, as these are more likely to be "requested". I don't know if this was considered, and I apologize if this is not the right place to suggest it.

I don't think that can/should be addressed in this PR. The matcher is blazing fast and the fs transversal is fully streaming/so the UI thread is not blocked. In your case the FS is simply too slow. There is not much we can do about that. Right now the FR transversal is single-threaded to allow the files to bu sorted alphabetically. We could add an option to disable sorting and run it in parallel but I doubt this would help much in your case since the bottleneck is likely IO/network and not CPU. If you are looking for files in the CWD you can use` <space>F` (an actually good improvement would be a breath first search but `ignore`/`walkdir` do not support that). Either way, all of these are not really related to the changes here.

dead10ck commented on helix-term/src/ui/mod.rs:
```diff
@@ -225,20 +209,27 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> Picker
-    .with_preview(|_editor, path| Some((path.clone().into(), None)))
+    .with_preview(|_editor, path| Some((path.clone().into(), None)));
+    let injector = picker.injector();
+    std::thread::spawn(move || {
+        for file in files {
+            if injector.push(file).is_err() {
+                break;
+            }
+        }
+    });
```
I've seen this in a couple places, where we're spinning up a dedicated thread for these operations. Why is this preferable to a tokio task? Is this just an effort to avoid interfering with the main event loop?

Are there any downsides to this approach? If I'm understanding this right, this is basically fire and forget, and if the thread errors or panics, there's nothing we can do, is that right?

pascalkuthe replied:
You are not supposed to block in async takes so tokio::spawn would be wrong to use here (it messes up tokio scheduling). We could use the blocking task spawner but in the best case that will reuse an existing sleeping thread (but in moat cases it would also just spawn a new thread itself). I could move to that but it ultimately doesn't really make a difference so I just reached for the standard library.

Ultimately for heavy compute and blocking io (you cant make readdir async the kernel doesnt support that) on a background thread that is throughput bound instead of waiting for a message or something tokio doesn't really offer any advantages (in fact the docs even recommknd rayon for that).

I guess it's true that we would miss panics here but I don't consider that a huge issue. A panic here means a panic in walkdir/ignore (it doesn't do much else). In the worst case you will stop seeing new files stream into the matcher.

dead10ck replied:
Makes sense, thanks for the explanation

dead10ck reviewed (dismissed):
Amazing work as always 🙂

pascalkuthe commented on helix-term/src/commands.rs:
```diff
@@ -2093,166 +2089,173 @@ fn global_search(cx: &mut Context) {
+                                let result = if let Some((_, doc)) = doc {
+                                    // there is already a buffer for this file
+                                    // search the buffer instead of the file because it's faster
+                                    // and captures new edits without requireing a save
```
This typo was actually already present on master (the code just moved) so I put it in seperate commit :D

cbr9 commented:
I have noticed that the RAM consumption is pretty high. Opening the file picker on my home directory takes about 1GB of RAM. When it's closed the memory is not freed, but when you reopen the picker, it rescans everything and takes another extra GB. @pascalkuthe

archseer commented:
Can you share how you made these measurements? If you're looking at `top` it's likely memory that's allocated to the application but not actually in use.

cbr9 commented:
I'm using [bottom](https://github.com/ClementTsang/bottom). It seems to be in use.

archseer commented:
I can't reproduce on my device. Closing the popup then reopening it several times keeps the RAM use at a constant size

archseer commented:
I still think bottom/top don't report this properly. Looking at `free -m` statistics before and after opening the picker a bunch of times gives me about identical values.

archseer commented:
That's only scoped to /proc/, I tested by opening under /nix/store

pascalkuthe commented:
not that @gabydd is also right here that a bug in the standard library has been fixed in 1.68 can also cause the crawler to become stuck which would lead to an actual memory leak. That is fixed just by compiling with a newer rustc version (we can probably raise the MSRV soon anyway)

pascalkuthe commented:
#8127 should fix this for the most part
